### ISSUE TITLE
[Bug] DSpark speculative decoding causes accuracy regression on DeepSeek-V4-Flash (AIME25 97.08→93.96)

### ISSUE THREAD

**Issue by fangtang1121** [sgl-project/sglang#32038]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

Enable DSpark speculative decoding on DeepSeek V4 Flash causes a stable accuracy regression compared to the non speculative baseline.

### Reproduction

DSpark:
python -m sglang.launch_server --model-path /media/disk1/deepseek-v4-flash-dspark --host 0.0.0.0 --port 30054 --tp 4 --moe-runner-backend flashinfer_mxfp4 --mem-fraction-static 0.85 --cuda-graph-max-bs 128 --max-running-requests 256 --trust-remote-code --moe-runner-backend flashinfer_mxfp4 --speculative-moe-runner-backend flashinfer_mxfp4 --speculative-algorithm DSPARK --speculative-eagle-topk 1 --speculative-num-draft-tokens 6

baseline:
python -m sglang.launch_server --model-path /media/disk1/deepseek-v4-flash-dspark --host 0.0.0.0 --port 30053 --tp 4 --moe-runner-backend flashinfer_mxfp4 --mem-fraction-static 0.9 --cuda-graph-max-bs 128 --max-running-requests 256 --trust-remote-code

### Environment

Python: 3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0]
CUDA available: True
GPU 0,1,2,3: NVIDIA B300 SXM6 AC
GPU 0,1,2,3 Compute Capability: 10.3
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 13.0, V13.0.88
CUDA Driver Version: 590.44.01
PyTorch: 2.11.0+cu130
sglang: 0.0.0.dev1+gd6ef68881
sglang-kernel: 0.4.5
flashinfer_python: 0.6.14
flashinfer_cubin: 0.6.14
flashinfer_jit_cache: 0.6.14+cu130
triton: 3.6.0
transformers: 5.12.1
torchao: 0.17.0+cu130
numpy: 2.3.5
aiohttp: 3.14.2
fastapi: 0.139.2
huggingface_hub: 1.24.0
interegular: 0.3.3
modelscope: 1.38.1
orjson: 3.11.9
outlines: 0.1.11
packaging: 26.2
psutil: 7.2.2
pydantic: 2.13.0
python-multipart: 0.0.32
pyzmq: 27.1.0
uvicorn: 0.51.0
uvloop: 0.22.1
vllm: Module Not Found
xgrammar: 0.2.1
openai: 2.6.1
tiktoken: 0.13.0
anthropic: 0.117.0
litellm: Module Not Found
torchcodec: 0.11.1+cu130
NVIDIA Topology:
        GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    NIC6    NIC7    NIC8    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NV18    NV18    NV18    NV18    NV18    NV18    NV18    PXB     NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     0-63,128-191    0               N/A
GPU1    NV18     X      NV18    NV18    NV18    NV18    NV18    NV18    NODE    PXB     NODE    NODE    SYS     SYS     SYS     SYS     SYS     0-63,128-191    0               N/A
GPU2    NV18    NV18     X      NV18    NV18    NV18    NV18    NV18    NODE    NODE    PXB     NODE    SYS     SYS     SYS     SYS     SYS     0-63,128-191    0               N/A
GPU3    NV18    NV18    NV18     X      NV18    NV18    NV18    NV18    NODE    NODE    NODE    PXB     SYS     SYS     SYS     SYS     SYS     0-63,128-191    0               N/A
GPU4    NV18    NV18    NV18    NV18     X      NV18    NV18    NV18    SYS     SYS     SYS     SYS     PXB     NODE    NODE    NODE    NODE    64-127,192-255  1               N/A
GPU5    NV18    NV18    NV18    NV18    NV18     X      NV18    NV18    SYS     SYS     SYS     SYS     NODE    PXB     NODE    NODE    NODE    64-127,192-255  1               N/A
GPU6    NV18    NV18    NV18    NV18    NV18    NV18     X      NV18    SYS     SYS     SYS     SYS     NODE    NODE    PXB     NODE    NODE    64-127,192-255  1               N/A
GPU7    NV18    NV18    NV18    NV18    NV18    NV18    NV18     X      SYS     SYS     SYS     SYS     NODE    NODE    NODE    PXB     NODE    64-127,192-255  1               N/A
NIC0    PXB     NODE    NODE    NODE    SYS     SYS     SYS     SYS      X      NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS
NIC1    NODE    PXB     NODE    NODE    SYS     SYS     SYS     SYS     NODE     X      NODE    NODE    SYS     SYS     SYS     SYS     SYS
NIC2    NODE    NODE    PXB     NODE    SYS     SYS     SYS     SYS     NODE    NODE     X      NODE    SYS     SYS     SYS     SYS     SYS
NIC3    NODE    NODE    NODE    PXB     SYS     SYS     SYS     SYS     NODE    NODE    NODE     X      SYS     SYS     SYS     SYS     SYS
NIC4    SYS     SYS     SYS     SYS     PXB     NODE    NODE    NODE    SYS     SYS     SYS     SYS      X      NODE    NODE    NODE    NODE
NIC5    SYS     SYS     SYS     SYS     NODE    PXB     NODE    NODE    SYS     SYS     SYS     SYS     NODE     X      NODE    NODE    NODE
NIC6    SYS     SYS     SYS     SYS     NODE    NODE    PXB     NODE    SYS     SYS     SYS     SYS     NODE    NODE     X      NODE    NODE
NIC7    SYS     SYS     SYS     SYS     NODE    NODE    NODE    PXB     SYS     SYS     SYS     SYS     NODE    NODE    NODE     X      NODE
NIC8    SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE     X

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_10
  NIC1: mlx5_11
  NIC2: mlx5_12
  NIC3: mlx5_13
  NIC4: mlx5_14
  NIC5: mlx5_15
  NIC6: mlx5_16
  NIC7: mlx5_17
  NIC8: mlx5_bond_0

ulimit soft: 1048576

**Comment by slchenchn** [comment]
This may be related to the verifier state rewrite bug fixed in my PR #32183. As a workaround, reducing ` --speculative-num-draft-tokens ` to below 4 may also avoid the issue.

**Comment by koolvn** [comment]
I should note that --speculative-num-draft-tokens=4 also works correctly. Problems begin when setting values > 4

**Comment by hassellof** [comment]
A datapoint that I think changes the workaround advice in this thread.

We hit DSpark output corruption on 4× RTX PRO 6000 (sm_120, TP=4 + DP attention + EP=4) with DeepSeek-V4-Flash-0731, and instead of testing two points we swept the whole depth range under an identical bursty agentic load (streaming, thinking on, 20-tool schemas). SGLang enforces `speculative_num_draft_tokens == gamma + 1`, so in terms of the flag used here:

| `--speculative-num-draft-tokens` | draft depth (gamma) | requests | corruption events |
|---|---|---|---|
| 4 | 3 | 10,885 (5 cold launches) | 0 |
| 5 | 4 | 11,040 (5 cold launches) | 0 |
| **6** | **5** | 2,148 inferred / 2,138 explicit | **10 / 7** |
| 7 | 6 | 2,282 | 0 |
| 8 | 7 | 2,198 | 0 |

**It is not monotonic.** 7 and 8 draft tokens are clean for us; only 6 is broken — draft depth 5, which is exactly `dspark_block_size` in the 0731 checkpoint config, and therefore the value you get whenever the depth is inferred instead of set. @fangtang1121's repro passes `--speculative-num-draft-tokens 6`, which lands precisely on it.

So "problems begin when setting values > 4" may be two samples either side of a single bad value rather than a ceiling. If the box is still available, the cheap check is to run **8**, not 5 — if 8 is clean there too, then "cap at 4" is the wrong lesson and everyone capping is giving up accept length for nothing.

@slchenchn's #32183 is real and we ship it as a backport — with the plan kernel instrumented we confirmed `verify_width` does reach `plan_compress_prefill` once it is applied — but it is necessary rather than sufficient: depth 5 still corrupts with that PR active.

Full evidence, repro protocol and the hypotheses we falsified by measurement (c4 ring capacity, verify window > `compress_ratio`, the `kMaxMTPDraftTokens = 4` row budget, NaN logits) are in #33800.

Caveat on transferring this: ours is sm_120 and yours is B300 sm_103, so I would treat "the checkpoint's native block size is special" as the shared hypothesis rather than assume an identical bug, until someone runs the sweep on datacentre Blackwell. Happy to run candidate patches on our hardware; each arm takes about 30 minutes.

Config, patch set and the measured numbers: https://github.com/ombori/deepseek-v4-flash-0731-sglang-4x-rtx-pro-6000

**Comment by apejcic** [comment]
Likely the same mechanism as #33800-on-B300. If your run uses the DSV4
backend with `speculative_num_draft_tokens > 4` (the DSpark checkpoint
default depth 5 → ndt=6), the compress write-plan under-persists the
verify window tail (hard-coded pad of 4), so post-accept compression
bakes stale ring slots into the state — a silent accuracy regression that
grows with output position, matching an AIME-style score drop rather than
a hard failure. #33872 predicts exactly this ndt>4 breakage and fixes it;
on B300 we measured the baseline ladder (ndt=4: 0%, ndt=6: ~42%, ndt=7:
56.5%, ndt=8: 75% corrupted requests on an echo probe) flatten to zero
anomalies across 1400 patched requests (d5/d6/d7, multiple cold boots).
If you can share your
`--speculative-*` flags we can say whether your config is in the affected
envelope; happy to share our repro protocol (opaque-id echo replay, much
cheaper than AIME for A/B-ing the fix).
